### PR TITLE
[QNN EP] Support quantized BatchNorm with per-channel DQ params on QNN HTP

### PR DESCRIPTION
**Pull request by qti-yuduo**
## Motivation:
QNN HTP was rejecting quantized BatchNorm models where parameters (scale, mean, var) come through DequantizeLinear nodes with per-channel INT8 quantization. This pattern is common in quantized models from quantization tools.

## Changes:

- Helpers to resolve BatchNorm params through DQ nodes to their underlying initializers
- Support per-channel dequantization for BatchNorm parameters
- Support input datatype of UFIXED_POINT_16
- Add unit test covering this QDQ params configuration

### CODE REVIEW

**Comment by tianleiwu** [comment]
From AI
----
## Summary
This PR enhances the QNN EP to support `BatchNorm` operators where the parameters (scale, mean, var) are supplied via DequantizeLinear (DQ) nodes with **per-channel INT8 quantization**. Previously, these configurations were rejected by the HTP backend.

## Key Changes
-   **Parameter Resolution**: Logic added to `batch_norm_op_builder.cc` to correctly trace back through DQ nodes to find the underlying initializers for batch norm parameters.
-   **Quantization Support**: Explicit support added for per-channel dequantization and `UFIXED_POINT_16` data types.
-   **Testing**: Added a dedicated unit test `BatchNormOpTests` covering this specific QDQ configuration.

## Review Analysis
### Correctness
-   **Gap Fill**: This addresses a specific limitation where valid quantized models (likely from tools like olive or AMCT) were failing on QNN HTP.
-   **Logic**: The code to unwrap the DQ params seems consistent with how other operators handle QDQ inputs.
-   **Selectors**: Updates to `qdq_selectors.cc` ensure the optimizer doesn't prematurely fuse or reject these nodes incorrectly.

### Performance
-   **Hardware Acceleration**: Enabling these nodes to run on the HTP (Hexagon Tensor Processor) instead of falling back to CPU (or failing) is a significant performance enabler for quantized models.

## Conclusion
This is a necessary fix for broad support of quantized models on Qualcomm hardware. The implementation includes necessary builder logic and verification tests. **LGTM**.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 1]
```diff
@@ -262,56 +263,70 @@ class BatchNormOpBuilder : public BaseOpBuilder {
     return Status::OK();
   }
 
+  // Maybe de-quantizes a 1D BatchNorm parameter tensor to double values.
+  Status MaybeDequantizeParamTensor(const TensorInfo& info,
+                                    const uint8_t* raw_ptr,
+                                    const size_t raw_ptr_length,
+                                    const char* tensor_name,
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: >const char* tensor_name [](http://example.com/codeflow?start=36&length=23)

POinters denote optional params and should be checked for nullptr. Otherwise, pass a reference to std::string or std::string_view #Resolved

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 2]
```diff
@@ -262,56 +263,70 @@ class BatchNormOpBuilder : public BaseOpBuilder {
...
+    const bool is_quantized = info.quant_param.IsQuantized();
+    const bool is_per_channel = info.quant_param.IsPerChannel();
+    const Qnn_QuantizeParams_t& quant_param = info.quant_param.Get();
+    if (is_per_channel) {
+      // Validate per-channel quantization parameters for 1D BatchNorm tensors.
+      // For 1D tensors, axis must be 0 and numScaleOffsets must be >= channel count.
+      ORT_RETURN_IF_NOT(quant_param.axisScaleOffsetEncoding.axis == 0,
+                        "Per-channel quantization axis must be 0 for 1D ", tensor_name, " tensor, got ",
+                        quant_param.axisScaleOffsetEncoding.axis);
+      ORT_RETURN_IF_NOT(quant_param.axisScaleOffsetEncoding.numScaleOffsets >= channel,
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: nit: why is this `>=` and not `==`? Wouldn't the number of scales/offsets match the channel dim?

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 3]
```diff
@@ -437,6 +433,67 @@ class BatchNormOpBuilder : public BaseOpBuilder {
...
+// Helper to resolve param initializer that may come from DQ nodes.
+Status ResolveParamInitializer(const QnnModelWrapper& qnn_model_wrapper,
+                               const NodeUnit& node_unit,
+                               const NodeUnitIODef& param,
+                               TensorInfo& tensor_info) {
+  const std::string& param_name = param.node_arg.Name();
+  if (tensor_info.is_initializer) {
+    return Status::OK();
+  }
+  for (const Node* dq_node : node_unit.GetDQNodes()) {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: Why is it necessary to manually look past the DQ nodes? I would have expected that the updates this PR makes to qdq_selectors.cc would make it so that checking the node unit's input directly would be sufficient.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 4]
```diff
@@ -262,56 +263,70 @@ class BatchNormOpBuilder : public BaseOpBuilder {
     return Status::OK();
   }
 
+  // Maybe de-quantizes a 1D BatchNorm parameter tensor to double values.
```
[the remark this answers is not in the document]
Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: Done

File: onnxruntime/test/providers/qnn/batch_norm_test.cc [hunk 5]
```diff
@@ -411,6 +413,104 @@ TEST_F(QnnHTPBackendTests, BatchNorm3D) {
...
+  return [input_def, scale_def, bias_def](ModelTestBuilder& builder,
+                                          std::vector<QuantParams<InputQType>>& output_qparams) {
+    const auto& input_shape = input_def.GetShape();
+    const auto& input_data = input_def.GetRawData();
+    const int64_t num_channels = input_shape[1];
+
+    // Input: float -> Q -> DQ (asymmetric for uint16 to match real models)
+    bool symmetric = sizeof(InputQType) == sizeof(uint16_t);
+    NodeArg* input = MakeTestInput(builder, input_def);
+    QuantParams<InputQType> input_qparams = GetTestInputQuantParams<InputQType>(input_def, symmetric);
```
[the remark this answers is not in the document]
Comment (qti-yuduo) [reply onnxruntime/test/providers/qnn/batch_norm_test.cc:433]: Fixed confusing comments.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc (hunk 1 above)
Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: Done.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc (hunk 2 above)
Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: Good catch. Done.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc (hunk 3 above)
Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: Removing.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 6]
```diff
@@ -437,6 +433,45 @@ class BatchNormOpBuilder : public BaseOpBuilder {
...
+    if (dq_node->OutputDefs()[0]->Name() == name) {
+      return qnn_model_wrapper.IsConstantInput(dq_node->InputDefs()[0]->Name());
+    }
+  }
+  return false;
+}
+
+// Adjust BatchNorm param types for QNN HTP compatibility.
+// Modifies scale/bias types in-place; quantization happens in Postprocess.
+void MaybeQuantizeAndOverrideParamType(Qnn_DataType_t x_dtype,
```
Comment (edgchen1) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: what does the "maybe quantize" part of the name mean? I don't see any quantization happening.

Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: There's dequantize, then requantize step (after preprocessing params, and quantize in postprocess). This method ensures we adjust parameter types when that re-quantize happen. Feel free to suggest better naming.

Comment (edgchen1) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: `OverrideParamTypeForRequantize`?

Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc]: Done.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 7]
```diff
@@ -437,6 +433,45 @@ class BatchNormOpBuilder : public BaseOpBuilder {
...
+
+// Helper to check if a BatchNorm param is constant - either direct initializer or through a DQ node.
+bool IsParamConstant(const QnnModelWrapper& qnn_model_wrapper,
+                     const NodeUnit& node_unit,
+                     const std::string& name) {
+  if (qnn_model_wrapper.IsConstantInput(name)) {
+    return true;
+  }
+  // Check if param comes through a DQ node with constant input
+  for (const Node* dq_node : node_unit.GetDQNodes()) {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:446]: nit: probably don't need to manually look past a DQ here either. can probably be cleaned up later.

Comment (qti-yuduo) [reply onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:446]: Will push a follow up.
